Source/JavaScriptCore/runtime/RegExpConstructor.cpp

```diff
@@ -83,18 +86,96 @@ RegExpConstructor::RegExpConstructor(VM& vm, Structure* structure)
     ASSERT(inherits(info()));
 
     putDirectWithoutTransition(vm, vm.propertyNames->prototype, regExpPrototype, PropertyAttribute::DontEnum | PropertyAttribute::DontDelete | PropertyAttribute::ReadOnly);
 
-    JSGlobalObject* globalObject = regExpPrototype->globalObject();
+    if (Options::useRegExpEscape())
+        JSC_NATIVE_FUNCTION_WITHOUT_TRANSITION("escape"_s, regExpConstructorEscape, static_cast<unsigned>(PropertyAttribute::DontEnum), 1, ImplementationVisibility::Public);
+
     GetterSetter* speciesGetterSetter = GetterSetter::create(vm, globalObject, JSFunction::create(vm, globalObject, 0, "get [Symbol.species]"_s, globalFuncSpeciesGetter, ImplementationVisibility::Public, SpeciesGetterIntrinsic), nullptr);
     putDirectNonIndexAccessorWithoutTransition(vm, vm.propertyNames->speciesSymbol, speciesGetterSetter, PropertyAttribute::Accessor | PropertyAttribute::ReadOnly | PropertyAttribute::DontEnum);
 }
 
+JSC_DEFINE_HOST_FUNCTION(regExpConstructorEscape, (JSGlobalObject* globalObject, CallFrame* callFrame))
+{
+    VM& vm = globalObject->vm();
+    auto scope = DECLARE_THROW_SCOPE(vm);
+
+    JSValue argValue = callFrame->argument(0);
+    if (!argValue.isString())
+        return throwVMTypeError(globalObject, scope, "RegExp.escape requires a string"_s);
+
+    JSString* argString = asString(argValue);
+    ASSERT(argString);
+
+    const String& string = argString->value(globalObject);
+    RETURN_IF_EXCEPTION(scope, { });
+
+    StringBuilder builder;
+    builder.reserveCapacity(string.length());
+
+    for (unsigned i = 0; i < string.length();) {
+        char32_t codePoint;
+        if (string.is8Bit())
+            codePoint = string.characters8()[i++];
+        else
+            U16_NEXT(string.characters16(), i, string.length(), codePoint);
+
+        if (builder.isEmpty() && isASCIIAlphanumeric(codePoint)) {
+            builder.append('\\', 'x', toStringWithRadix(codePoint, 16));
+            continue;
+        }
+
+        if (isASCII(codePoint) && strchr("^$\\.*+?()[]{}|/", codePoint)) {
+            builder.append('\\', codePoint);
+            continue;
+        }
+
+        switch (codePoint) {
+        case '\t':
+            builder.append('\\', 't');
+            continue;
+        case '\n':
+            builder.append('\\', 'n');
+            continue;
+        case '\v':
+            builder.append('\\', 'v');
+            continue;
+        case '\f':
+            builder.append('\\', 'f');
+            continue;
+        case '\r':
+            builder.append('\\', 'r');
+            continue;
+        default:
+            break;
+        }
+
+        if ((isASCII(codePoint) && strchr(",-=<>#&!%:;@~'`\"", codePoint)) || isStrWhiteSpace(codePoint) || U16_IS_SURROGATE(codePoint)) {
+            if (isLatin1(codePoint))
+                builder.append('\\', 'x', pad('0', 2, toStringWithRadix(codePoint, 16)));
+            else if (U_IS_BMP(codePoint))
```

Comment:
Here and below, I wonder if it would be more consistent to say `!U16_IS_SURROGATE`?

Reply:
is there a specific reason for using `U16_IS_SURROGATE` instead of `U_IS_BMP`?  i was trying to follow what `stringFromCodePoint` does.

also technically i think all we really care about is `codePoint <= 0xFFFF` according to the spec <[https://tc39.es/ecma262/#sec-unicodeescape](https://tc39.es/ecma262/#sec-unicodeescape)>

Reply:
What I mean is that you're in the implicit "else" after [step 5a](https://bakkot.github.io/proposal-regex-escaping/#sec-encode), so having ruled out the Latin1 case, it would seem most natural to describe the follow-up question as "surrogate pair or not surrogate pair?".

(I assume that the code should work correctly either way, so it should just be a matter of readability.)

Reply:
i dont really know enough about the intended use to comment so i'll trust your knowledge about it, but the spec specifically says to check for `<= 0xFFFF` ([step 2 of `UTF16EncodeCodePoint`](https://tc39.es/ecma262/#sec-utf16encodecodepoint))

also i think its worth mentioning that im not sure the `else` will ever actually even be reachable given the parent `if` above since i think all the (current) values of `isStrWhiteSpace`/`U16_IS_SURROGATE`/etc. are `<= 0xFFFF`

maybe im doing the bitwise math incorrectly, but if we did use `!U16_IS_SURROGATE` then wouldn't something like `RegExp.escape("0xd800")` result in something like `"\\ud7f6\\udc00"` instead of just `"\\ud800"`?

Reply:
So I was saying something rather simpler, based on the way that you've written your `if` here—either it's some specific ASCII chars, or it's `isStrWhiteSpace`, or it's `U16_IS_SURROGATE`. But I was forgetting that `isStrWhiteSpace` includes a handful of non-ASCII whitespace/line terminator characters. So ultimately I think what you've got makes sense.